File: PA1/List.java
package pa1;

public class List extends Lex {

	private class Node {
		private int item;
		private Node next;
		private Node pre;

		Node(int data) {
			this.item = data;
			next = null;
			pre = null;
		}

		Node getPre() {
			return pre;
		}

		Node getNext() {
			return next;
		}

		void setPre(Node node) {
			pre = node;
		}

		void setNext(Node node) {
			next = node;
		}


		int getItem() {
			return item;
		}
	}

	public <T> void cout(T x) // / Easier to print out stuff this way
	{
		System.out.println(x);
	}

	private Node head = null;
	private Node back = null;
	private Node cursor = null;
	private int length = 0;
	private int index = -1;

	// Constructor
	List()// Creates a new empty list.
	{
//		cout("Default Constructor Initiated x1");
	};

	// Access functions
	int length() {
		// Returns the number of elements in this List.
		return length;
	};

	int index() {
		cout("index():: print cursor : " + cursor );
		if (cursor != null)
			return index;
		else
			return -1;
	};// If cursor is defined, returns the index of the cursor element,
		// otherwise returns -1.

	int front() {
		return head.getItem();
	}; // Returns front element. Pre: length()>0

	int back() {
		return back.getItem();
	}; // Returns back element. Pre: length()>0

	int get() {
		if (length() <= 0 || index() < 0) {
			cout("get()::length() <= 0 || index() < 0");
			return -1;
		}
		return cursor.getItem();
	}; // Returns cursor element. Pre: length()>0, index()>=0

	boolean equals(List L) {
		// Returns true if and only if this List and L are the same
		// integer sequence. The states of the cursors in the two Lists
		// are not used in determining equality.
		return false;
	}

	// Manipulation procedures:
	void clear() {
		// Resets this List to its original empty state.
		head = null;
		back = null;
		cursor = null;
		index = -1;
		length = 0;
	}

	void moveFront() {
		// If List is non-empty, places the cursor under the front element,
		// otherwise does nothing.
		if (length == 0) {
			// throw new RuntimeException(
			// "List is Empty");
			//
			cout("moveFront: Error: Empty List");
			return;
		}
		cout("cursor = head now");
		cursor = head;
		index = 0;
	}

	void moveBack() {
		// If List is non-empty, places the cursor under the back element,
		// otherwise does nothing.
		if (length == 0) {
			// throw new RuntimeException(
			// "List is Empty");
			//
			cout("moveBack: Error: Empty List");
			return;
		}
		//cout("cursor = head now");
		cursor = back;
		index = length - 1;
	}

	void movePrev() {
		// If cursor is defined and not at front, moves cursor one step toward
		// front of this List, if cursor is defined and at front, cursor becomes
		// undefined, if cursor is undefined does nothing.
		if (cursor == null) {

			cout("cursor == null");
			return;
		}
		cursor = cursor.getPre();
		index--;
//		cout("cursor != null and index: " + index + " cursor : " + cursor);

	}

	void moveNext() {
		// If cursor is defined and not at back, moves cursor one step toward
		// back of this List, if cursor is defined and at back, cursor becomes
		// undefined, if cursor is undefined does nothing.
		if (cursor == null) {
			return;
		}
		cursor = cursor.next;
		index++;

	}

	void prepend(int data) {
		// Insert new element into this List. If List is non-empty,
		// insertion takes place before front element.
		Node node = new Node(data);
		if (length == 0) {
			head = node;
			back = node;
			length++;
		} else {
			head.setPre(node);
			node.setNext(head);
			head = node;
			length++;
		}
	}

	void append(int data) {
		// Insert new element into this List. If List is non-empty,
		// insertion takes place after back element.
		Node node = new Node(data);
		if (length() == 0) {
			head = node;
			back = node;
			length++;
		} else {
			back.setNext(node);
			node.setPre(back);
			back = node;
			length++;
		}
	}

	void insertBefore(int data) {
		// Insert new element before cursor.
		// Pre: length()>0, index()>=0
		if (length() <= 0 || index() < 0) {
			cout("length() <= 0 || index() < 0");
			return;
		}
		Node node = new Node(data);
		if (cursor.getPre() != null) // in case the cursor element is the head
		{
			node.setPre(cursor.getPre());
			cursor.getPre().setNext(node);
		} else {
			head = node;
		}
		cursor.setPre(node);
		node.setNext(cursor);
		length++;
	}

	void insertAfter(int data) {
		// Inserts new element after cursor.
		// Pre: length()>0, index()>=0
		if (length() <= 0 || index() < 0) {
			cout("length() <= 0 || index() < 0");
			return;
		}
		Node node = new Node(data);
		if (cursor.getNext() != null) {
			node.setNext(cursor.getNext());
			cursor.getNext().setPre(node);
		} else {
			back = node;
		}
		cursor.setNext(node);
		node.setPre(cursor);
		length++;

	}

	void deleteFront() {
		// Deletes the front element. Pre: length()>0
		if (length > 0) {
			head = head.getNext();
			length--;
		}
	}

	void deleteBack() {
		// Deletes the back element. Pre: length()>0
		if (length > 0) {
			back = back.getPre();
			length--;
		}
	}

	void delete() {
		// Deletes cursor element, making cursor undefined.
		// Pre: length()>0, index()>=0
		if (length() <= 0 || index() < 0) {
			cout("length() <= 0 || index() < 0");
			return;
		}

		cursor = null;
		index = -1;
	}
	
	void print()  // use to test list, delete later.
	{
		if(length() == 0)
			return;
		Node temp = head;
		while(temp != null)
		{
			cout(temp.getItem());
			temp = temp.getNext();
		}
	}

}

File: PA2/List/List.c
//-----------------------------------------------------------------------------
// Tyler Hu
// #1597112
// Programming assignment 2
// List.h
// List ADT Implementation
//-----------------------------------------------------------------------------

#include <stdlib.h>
#include <stdio.h>
#include "List.h"

// STRUCTS --------------------------------------------------------------------
// private NodeObj type
typedef struct NodeObj
{
    int data;
    struct NodeObj* prev;
    struct NodeObj* next;
} NodeObj;

// private Node type
//This will allow future assignemnts to just being Node variable name
//otherwise you'll have to type struct NodeObj *front every single time
typedef NodeObj* Node;

// private List type
typedef struct ListObj
{
    Node front;
    Node back;
    Node cursor;
    int length,index;
} ListObj;


// CONSTRUCTOR-DESTRUCTORS ---------------------------------------------------
// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data)
{
    Node node = malloc(sizeof(NodeObj));
    node->data = data;
    node->prev = NULL;
    node->next = NULL;
    return(node);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN)
{
    if( pN!=NULL && *pN!=NULL )
    {
        (*pN)->next = (*pN)->prev = NULL;
        free(*pN);
        *pN = NULL;
    }
}

// newList()
// Returns reference to new empty Queue object.
List newList(void)
{
    List L;
    L = malloc(sizeof(ListObj));
    L->front = L->back = L->cursor = NULL;
    L->length = 0;
    L->index = -1;
    return(L);
}
// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.S


int isEmpty(List L)
{
    if( L==NULL )
    {
        printf("List Error: calling isEmpty() on NULL List reference\n");
        exit(1);
    }
    return(L->length==0);
}

void freeList(List* L)
{
    if(*L == NULL)
        return;

    if(length(*L) == 0)
        return;

    clear(*L);
    free(*L);
    *L = NULL;

}

void clear(List L)
{
    moveFront(L);
    while(length(L)>0)
    {
        deleteFront(L);
        moveFront(L);
    }

}

void deleteFront(List L)  // Deletes element at front of L
{
    if( L==NULL || length(L) == 0)
    {
        return;
    }

    Node N = NULL;
    N = L->front;
    if(L->front == L->cursor)
    {
        L->cursor =  NULL;
        L->index = -1;
    }
    if(length(L) > 1)
    {
        L->front = L->front->next;
        L->index--;
    }
    else
    {
        L->front = L->back = L->cursor =  NULL;
    }
    L->length--;
    freeNode(&N);
}
//ACCESS FUNCTIONS ----------------------------------------------------------

// length()
// Returns true (1) if L is empty, otherwise returns false (0)


int get(List L) // Returns cursor element. Pre: length()>0, index()>=0
{
    if (L->length <= 0 || index(L) < 0)
    {
//			cout("get()::length() <= 0 || index() < 0");
        return -1;
    }
    return(L->cursor->data);
}

int front(List L) // Returns front element. Pre: length()>0
{
    if( L==NULL )
    {
        printf("Lueue Error: calling getFront() on NULL Lueue reference\n");
        exit(1);
    }
    if( length(L) == 0)
    {
        printf("Lueue Error: calling getFront() on an empty Lueue\n");
        exit(1);
    }
    return(L->front->data);
}

int back(List L) // Returns rear element. Pre: length()>0
{
    if( L==NULL )
    {
        printf("Lueue Error: calling getFront() on NULL Lueue reference\n");
        exit(1);
    }
    if(length(L) == 0 )
    {
        printf("Lueue Error: calling getFront() on an empty Lueue\n");
        exit(1);
    }

    return(L->back->data);
}

int index(List L) //If cursor is defined, returns the index of the cursor element
{
    if(L->cursor != NULL)
    {
        return (L->index);
    }
    return -1; //if cursor is pointed towards NULL
}

int length(List L) // Returns the length of L.
{
    if( L==NULL )
    {
        printf("List Error: calling length() on NULL List reference\n");
        return 0;
    }
    return(L->length);
}

int equals(List A, List B)
{
    moveFront(A);
    moveFront(B);

    while(index(A) >= 0 || index(B) >= 0)
    {
        if(get(A) != get(B))
        {
            return 0;
        }
        moveNext(A);
        moveNext(B);
    }
    return 1;
}

//MANIPULATORS -----------------------------------------------------------------

void moveFront(List L) //move cursor to the front
{
    if (L->length == 0)
    {
        return;
    }
    L->cursor = L->front;
    L->index = 0;
}

void moveBack(List L) //move cursor to the front
{
    if (L->length == 0)
    {
        return;
    }
    L->cursor = L->back;
    L->index = L->length -1 ;
}



void deleteBack(List L)
{
    if( L==NULL || length(L) == 0)
    {
        return;
//        printf("List Error: calling deleteFront() on NULL List or length() reference\n");
//        exit(1);
    }

    Node N = NULL;
    N = L->back;
    if(L->back == L->cursor)
    {

        L->cursor = 0;
    }
//    }
    if( length(L) >1 )
    {
        L->back = L->back->prev;
        L->back->next = NULL;
    }
    else
    {
        L->front = L->back = NULL;
    }
//    printf("N pointer: %i \n", N);
    freeNode(&N);
//    printf("N pointer: %i \n", N);
    L->length--;
}
// If cursor is defined and not at back, moves cursor one step toward
// back of this List, if cursor is defined and at back, cursor becomes
// undefined, if cursor is undefined does nothing.
void append(List L, int data) // Places new data element at the start of List
{

    if( L==NULL )
    {
        printf("List Error: calling append() on NULL List reference\n");
        return;
    }
    Node N = newNode(data);
    if( length(L) == 0)
    {
        L->front = L->back = N;

    }
    else
    {

        N->prev = L->back;
        L->back->next = N;
        L->back = N;
    }
    L->length++;
}

void moveNext(List L)
{
    // If cursor is defined and not at back, moves cursor one step toward
    // back of this List, if cursor is defined and at back, cursor becomes
    // undefined, if cursor is undefined does nothing.
    if (L->cursor == NULL)
    {
        return;
    }

    L->cursor = L->cursor->next;
    L->index++;
}

void movePrev(List L)
{
    // If cursor is defined and not at front, moves cursor one step toward
    // front of this List, if cursor is defined and at front, cursor becomes
    // undefined, if cursor is undefined does nothing.
    if (L->cursor == NULL)
    {
        return;
    }
    L->cursor = L->cursor->prev;
    L->index--;
//		cout("cursor != null and index: " + index + " cursor : " + cursor);

}

void prepend(List L, int data) // Places new data element at the start of List
{
    if( L==NULL )
    {
        printf("List Error: calling append() on NULL List reference\n");
        return;
    }

    Node N = newNode(data);
    if( length(L) == 0 )
    {
        L->front = L->back = N;
    }
    else
    {

        L->front->prev = N;
        N->next = L->front;
        N->prev = NULL;
        L->front = N;
        L->index++;
    }
    L->length++;
}

void insertBefore(List L, int data)
{
    if (length(L) <= 0 || index(L) < 0)
    {
        return;
    }
    if(L->cursor == L->front)
    {
        prepend(L,data);
        return;
    }
    Node N = newNode(data);
    N->prev = L->cursor->prev;
    L->cursor->prev->next = N;
    L->cursor->prev = N;
    N->next = L->cursor;
    L->length++;
    L->index++;
}



void insertAfter(List L, int data)
{
    if (length(L) <= 0 || index(L) < 0)
    {
        return;
    }
    if(L->cursor == L->back)
    {
        append(L,data);
        return;
    }
    Node N = newNode(data);
    N->next= L->cursor->next;
    L->cursor->next->prev = N;
    N->prev = L->cursor;
    L->cursor->next = N;
    L->length++;
}
void printList(FILE* out, List L)
{
    Node N = NULL;
    if( L==NULL )
    {
        printf("List Error: calling printList() on NULL List reference\n");
        return;
    }
    moveFront(L);
    for(N = L->front; N != NULL; N = N->next)
    {
//        printf("%i", N->data);
        fprintf(out, "%i", N->data);
        if(index(L)!= (length(L)-1))
        {
//            printf("\n %i <- index length-1 -> %i \n",index(L), (length(L)-1));
            fprintf(out, " ");
//            printf(" ");
        }
        moveNext(L);
    }
}

void delete(List L)
{
    if(L->cursor == NULL)
    {
        return;
    }
    if(L->cursor == L->front)
    {
        L->index = -1;
        deleteFront(L);
        return;
    }
    if(L->cursor == L->back)
    {
        L->index = -1;
        deleteBack(L);
        return;
    }
    L->length--;
    L->cursor->prev->next = L->cursor->next;
    L->cursor->next->prev = L->cursor->prev;
    freeNode(&L->cursor);
    L->index = -1;
}

List copyList(List L)
{
    if(L == NULL)
    {
        exit(1);
    }
    List A = newList();
    Node temp = L->cursor;
    int tempIndex = L->index;
    moveFront(L);
    while(index(L)>=0)
    {
        append(A, get(L));
        moveNext(L);
    }
    L->cursor = temp;
    L->index = tempIndex;
    return (A);
}
